src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -1524,8 +1533,10 @@ _mongoc_topology_push_server_session (mongoc_topology_t *topology,
    }
 
    /* If session is expiring or "dirty" (a network error occurred on it), do not
-    * return it to the pool. */
-   if (_mongoc_server_session_timed_out (server_session, timeout) ||
+    * return it to the pool. Sessions do not expire when the topology type is
+    * load balanced. */
+   if ((!loadbalanced &&
+        _mongoc_server_session_timed_out (server_session, timeout)) ||
```

Comment:
Given how frequently you combine logic for `loadbalanced` and `_mongoc_server_session_timed_out`, would it make sense to have `_mongoc_server_session_timed_out` aware of the topology and just return `false` on its own for `loadbalanced` topologies?

I'm curious if `_mongoc_server_session_timed_out` is called from anywhere else that might need to care about this.

On the flip side, I understand doing so might violate a separation of concerns if sessions otherwise have no reason to know about the topology in which they were created.

Reply:
I would prefer avoiding any circular dependency. The topology maintains the session pool, and the client session objects do not know about the topology object. `mongoc-topology-private.h` includes `mongoc-client-session-private.h`.